weak_supervision/state_machines/transition_functions/basic_transition_function.py

```diff
@@ -0,0 +1,421 @@
+                    group_curr_log_probs.append(curr_log_probs)
+
+
+                log_probs = torch.cat(group_log_probs, dim=0)
+                curr_log_probs = torch.cat(group_curr_log_probs, dim=0)
+                action_embeddings = torch.cat(group_action_embeddings, dim=0)
+                log_probs_cpu = log_probs.data.cpu().numpy().tolist()
+                batch_states = [(log_probs_cpu[i],
+                                 group_indices[i],
+                                 log_probs[i],
+                                 action_embeddings[i],
+                                 group_actions[i])
+                                for i in range(len(group_actions))
+                                if (not allowed_actions or
+                                    group_actions[i] in allowed_actions[group_indices[i]])]
+                # this is the only change required to sample from current log probabilities instead of beam search
+                if sample:
+                    categorical_dist = Categorical(logits=curr_log_probs)
+                    _, group_index, log_prob, action_embedding, action = batch_states[categorical_dist.sample()]
+                    new_states.append(make_state(group_index, action, log_prob, action_embedding))
+
+                else:
+                    batch_states.sort(key=lambda x: x[0], reverse=True)
+                    if max_actions:
+                        batch_states = batch_states[:max_actions]
+                    for _, group_index, log_prob, action_embedding, action in batch_states:
+                        new_states.append(make_state(group_index, action, log_prob, action_embedding))
+
+        return new_states
+
+    def _take_first_step(self,
+                         state: GrammarBasedState,
+                         sample_states: bool,
+                         allowed_actions: List[Set[int]] = None) -> List[GrammarBasedState]:
+        # We'll just do a projection from the current hidden state (which was initialized with the
+        # final encoder output) to the number of start actions that we have, normalize those
+        # logits, and use that as our score.  We end up duplicating some of the logic from
+        # `_compute_new_states` here, but we do things slightly differently, and it's easier to
+        # just copy the parts we need than to try to re-use that code.
+
+        # (group_size, hidden_dim)
+        hidden_state = torch.stack([rnn_state.hidden_state for rnn_state in state.rnn_state])
+        # (group_size, num_start_type)
+        start_action_logits = self._start_type_predictor(hidden_state)
+        log_probs = torch.nn.functional.log_softmax(start_action_logits, dim=-1)
+
+        sorted_log_probs, all_actions = log_probs.sort(dim=-1, descending=True)
+        all_actions = all_actions.detach().cpu().numpy().tolist()
+        if sample_states:
+            # (group_size,) one action per group element
+            sampler = Categorical(logits=sorted_log_probs)
```

Comment:
Why do you have to use sorted `log_probs` here? You could have just passed `log_probs`, correct?

Reply:
so I wanted sampler to be able to return indices into all_actions which works consistently only if we use sorted_log_probs